File: 0295. Find Median from Data Stream/Solution.py
import heapq


class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min_heap = []
        self.max_heap = []

    def addNum(self, num: int) -> None:
        if self.max_heap and num <= -self.max_heap[0]:
            heapq.heappush(self.max_heap, -num)
            if len(self.max_heap) > len(self.min_heap) + 1:
                heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        else:
            heapq.heappush(self.min_heap, num)
            if len(self.min_heap) > len(self.max_heap) + 1:
                heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))

    def findMedian(self) -> float:
        if len(self.max_heap) > len(self.min_heap):
            return -self.max_heap[0]
        elif len(self.max_heap) < len(self.min_heap):
            return self.min_heap[0]
        else:
            return -self.max_heap[0] * 0.5 + self.min_heap[0] * 0.5


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295. Find Median from Data Stream/Solution.cpp
#include <functional>
#include <queue>
using namespace std;

class MedianFinder {
public:
    MedianFinder()
    {
    }

    void addNum(int num)
    {
        if (!max_heap.empty() && num <= max_heap.top()) {
            max_heap.emplace(num);
            if (max_heap.size() > min_heap.size() + 1) {
                min_heap.emplace(max_heap.top());
                max_heap.pop();
            }
        } else {
            min_heap.emplace(num);
            if (min_heap.size() > max_heap.size() + 1) {
                max_heap.emplace(min_heap.top());
                min_heap.pop();
            }
        }
    }

    double findMedian()
    {
        if (min_heap.size() > max_heap.size()) {
            return min_heap.top();
        } else if (max_heap.size() > min_heap.size()) {
            return max_heap.top();
        } else {
            return max_heap.top() * 0.5 + min_heap.top() * 0.5;
        }
    }

private:
    priority_queue<int, vector<int>, greater<int>> min_heap;
    priority_queue<int> max_heap;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
